File: src/core/legacy/ga/cross/TwoPointCrossover.js

'use strict';

const utils = require('../../utils')
  , Crossover = require('./Crossover');

class TwoPointCrossover extends Crossover {

  _crossover(pr1, pr2) {
    const pos1 = utils.math.randomInt(0, pr1.length - 1)
      , pos2 = utils.math.randomInt(0, pr1.length - 1)
      , to = Math.max(pos1, pos2);

    for (let i = Math.min(pos1, pos2); i <= to; i += 1) {
      let tmp = pr1[i];
      pr1[i] = pr2[i];
      pr2[i] = tmp;
    }
  }

}

module.exports = TwoPointCrossover;

File: src/core/main/java/workschedule/genetics/crossover/TwoPointCrossover.java
package workschedule.genetics.crossover;

import workschedule.genetics.Chromosome;
import workschedule.utils.MathUtils;
import workschedule.utils.Pair;

public final class TwoPointCrossover extends AbstractCrossover {
  @Override
  protected void implementation(final Pair<Chromosome> parents) {
    int pos1 = MathUtils.randomInt(0, parents.one.getLength() - 1);
    int pos2 = MathUtils.randomInt(0, parents.one.getLength() - 1);

    int from = Math.min(pos1, pos2);
    int to = Math.max(pos1, pos2);

    for (int i = from; i <= to; i += 1) {
      boolean tmp = parents.one.getGene(i);
      parents.one.setGene(i, parents.two.getGene(i));
      parents.two.setGene(i, tmp);
    }
  }
}
